doc/developer/design/20230411_envelope_upsert_order_by_timestamp.md

```diff
@@ -0,0 +1,110 @@
+- Feature name: Envelope upsert with order by timestamp (limited scope)
+- Associated: https://github.com/MaterializeInc/materialize/issues/16512
+
+# Summary
+
+- Support user specified order by in Kafka upsert sources in a limited capacity. If users want to specify custom ordering which is semantically meaningful to them, there's currently no easy way to do that. They could potentially use `ENVELOPE NONE` along with a `DISTINCT ON` view to get similar results, but it's cumbersome and lacks the automatic compaction on keys with normal `ENVELOPE UPSERT`.
+- Currently, will limit the scope to allow only `ORDER BY` the `TIMESTAMP` and the `OFFSET` metadata fields. Depending upon user feedback this can later be expanded to support expressions and extractions from other columns like `HEADERS`.
+
+# Scope
+
+The scope will be limited to only allow `ORDER BY ( TIMESTAMP, OFFSET )` (ascending order).
+
+- In a key-compacted Kafka topic, value = NULL is taken to mean that it’s a delete tombstone and that the corresponding key should be removed. If we want to borrow that tombstone semantic for our more flexible upsert source, it would make sense to limit the `ORDER BY` columns to only the metadata columns.
+- Limiting it to only `TIMESTAMP` and `OFFSET` because
+    - `OFFSET` is already default. We are making it explicit here and it's going to be required that the user always mention `OFFSET` as one of the columns in ORDER BY.
+    - `PARTITION`, `KEY`, `TOPIC`, would not make for meaningful orderings:
+        - The `PARTITION` of a record is typically derived from its key, and so is the same for every record with a given key.
+        - `KEY` is, by definition, the same for every record with a given key.
+        - `TOPIC` is the same for every record in the topic.
+    - `HEADERS` by itself would not also make a good order by clause (it’s a json blob), but later we can probably allow expressions so that values could be extracted from it.
+- Limiting to only ascending order because no users have expressed a need for descending.
+
+# Explanation
+
+## Proposal
+
+Make it possible with unsafe-mode to provide custom ORDER BY clause which will override the default ordering by offset. The given ORDER BY identifiers should refer to only the included TIMESTAMP and OFFSET metadata columns. It will be required that the user always mention OFFSET as one of the columns for the ORDER BY clause.
+
+The option will be part of the `ENVELOPE UPSERT` clause with the following grammar:
+
+`ENVELOPE UPSERT [(ORDER BY (<expr>) [ASC])]`
+
+The `ASC` modifier is optional noise for specifying ascending ordering, for symmetry with the `ORDER BY` clause in `SELECT` statements.
+
+Examples of valid syntax and semantics:
+- `CREATE SOURCE ... INCLUDE TIMESTAMP ENVELOPE UPSERT ( ORDER BY ( TIMESTAMP, OFFSET ) ASC )`
```

Comment:
```
CREATE SOURCE ... INCLUDE TIMESTAMP ENVELOPE UPSERT ( ORDER BY ( TIMESTAMP  ASC, OFFSET ASC) )
```

Reply:
I will update the design doc in follow up implementation PR https://github.com/MaterializeInc/materialize/pull/18567

Reply:
Btw, I was trying out select statements in materialize. Parentheses around expressions with ordering is erroring.

```
 select * from texttext order by (key ASC, text DESC);
ERROR:  Expected right parenthesis, found ASC
LINE 1: select * from texttext order by (key ASC, text DESC);
```

The following works though
```
select * from texttext order by key ASC, text;

select * from texttext order by (key, text) ASC;
```

Reply:
Should we lose the parentheses around the multiple order by statements with order then? Or go back to limiting it to one tuple expression with optional ASC?

Reply:
Right! The syntax for the `ORDER BY` clause in `SELECT` is:

```
select_order_by ::= ORDER BY <order_by_col> [, <order_by_col>]*
order_by_col := <expr> [ASC | DESC]
expr := <ident> | ( <expr> ) | ...
```

Whereas the syntax for the `ORDER BY` option in `CREATE SOURCE` will be:

```
create_source_option_order_by ::= ORDER BY (<order_by_col> [, <order_by_col>]*)
order_by_col := <expr> [ASC | DESC]
expr := <ident> | ( <expr> ) | ...
```

So the syntaxes have to _slightly_ deviate, because in `CREATE SOURCE` we need parens around the order by columns in order to disambiguate between a new `<order_by_col>` and a new `CREATE SOURCE` option.

The equivalent of

```
select * from texttext order by (key, text) ASC;
```

would be:

```
CREATE SOURCE ... ENVELOPE UPSERT (ORDER BY ((key, text) ASC))
```

Reply:
Makes sense. Thanks for explaining that!